Updates documentation of build process

conorsch opened the pull request:
Overhauls the documentation to reference recent changes (#105, #106), with an eye toward accessibility. In particular, strives to make the rebuild-from-source workflow clear and concise. 

Although the diff is complicated, only one file should matter to reviewers: `docs/build.md`. Consult that file for the latest documentation on rebuilding kernels from source packages. If anything is unclear, comment here, and we'll revise prior to merge.

Closes #102.

conorsch committed:
Updates SecureDrop kernel build docs

As often happens with technical documentation, the build guide for
SecureDrop kernels was out of date. Two major problems stood out:

  1. FPF staff with access to grsecurity portals couldn't follow
     the steps (recent PRs to this repo have adjusted the flow).
  2. SecureDrop Admins wishing to compile a kernel image from the
     source package would have a hard time reconstructing the steps.

Point 1) was rather easy to solve, given that the recent PRs vastly
simplified the flow. Docs are now accurate and should work for
subsequent kernel builds the next time we perform an update for
SecureDrop servers.

Point 2) is a bit trickier, because we're trying to provide a solid
documentation flow for a use case we haven't ever directly addressed
before: Admins wanting to recompile the kernel from source to verify
that it is correct (or simply to exercise their rights under the GPL).
Will need to ping some community members for review here to validate
that the docs are both clear and adequate.

conorsch committed:
Creates dedicated `grsec-rebuild-securedrop` VM

For SecureDrop we build and retain linux-source packages, which include
a tarball of the entire kernel source immediately prior to configuration
and compilation. Anyone trying to use that source tarball to rebuild a
kernel image package will need a few things:

  * build environment (apt dependencies and such)
  * tarball of source (requested from FPF)
  * ubuntu-overlay

The Ubuntu Trusty kernel overlay is *not* included in the source
tarball, so we much clone that separately. The recommended command in
the rebuild playbook explicitly states that the overlay should be
applied via CLI flags.

In order to achieve reuse of the grsec build logic, without attempting
to fetch the restricted-access grsecurity patches, added a few tags to
the main task list so that the tasks can be selectively included. That
way, anyone who wants to rebuild the kernel image from source does not
need authentication credentials for grsecurity. Instead, everything
required will be provided by the playbook and docs.

Also tweaks the `grsec-install` machine to use the same Trusty base
image as all the other Trusty boxes, for consistency in testing.

conorsch committed:
Overhauls README documentation

Recent PRs have revised the main logic flow, requiring substantial
updates to the documentation. Of particular note is the lack of "test"
patches due to discontinuation, and the recent hard requirement for
grsecurity login credentials in order to access patches (regardless of
patch type). Since the build documentation was separately updated,
the intro to the README now references it explicitly, because it may
be useful to SecureDrop Admins researching the build process.

conorsch commented:
Ping @dachary for review! See `docs/build.md` in this branch.

ghost commented:
will do :-)

dachary commented on docs/build.md:
```diff
@@ -1,246 +1,74 @@
+package from the original source code. To do so, first request the
+`linux-source-*.deb` package from FPF as described in the
+[SecureDrop source offer]. Once you have obtained that file, follow
+the instructions below.
```
I assume these instructions will work on a pristine Ubuntu 14.04, right ?

dachary replied:
nevermind, I was not looking at the right place

ghost commented:
For some reason **hkps.pool.sks-keyservers.net** failed (timed out) and **pgp.mit.edu**. Also the expected name is different.

<pre>
modified   roles/build-grsec-kernel/defaults/main.yml
@@ -27,7 +27,7 @@ grsecurity_build_download_directory: "{{ ansible_env.HOME }}/linux"
 grsecurity_build_linux_source_directory: >-
   {{ grsecurity_build_download_directory }}/linux-{{ linux_kernel_version }}
 
-grsecurity_build_gpg_keyserver: hkps.pool.sks-keyservers.net
+grsecurity_build_gpg_keyserver: pgp.mit.edu
 
 # Assumes 64-bit (not reading machine architecture dynamically.)
 grsecurity_build_deb_package: >-
@@ -72,7 +72,7 @@ grsecurity_build_download_password: "{{ lookup('env', 'GRSECURITY_PASSWORD')|def
 
 # List of GPG keys required for building grsecurity-patched kernel.
 grsecurity_build_gpg_keys:
-  - name: Greg Kroah-Hartman GPG key (Linux stable release signing key)
+  - name: Greg Kroah-Hartman (Linux kernel stable release signing key)
     fingerprint: 647F28654894E3BD457199BE38DBBDC86092693E
   - name: kernel.org checksum autosigner GPG key
     fingerprint: B8868C80BA62A1FFFAF5FDA9632D3A06589DA6B1
</pre>

conorsch commented:
@dachary Thanks for reporting, I've definitely had that trouble before (@msheiny too), so we should just update the default keyserver in the role to use pgp.mit.edu. (If you were saying that pgp.mit.edu _also_ failed for you, please elaborate!)

As for the `name:` parameters on the key info dicts, those are just for human benefit, so that it's possible to read the list of fingerprints and figure out who's who—the `name` parameters are _not_ used programmatically during the key fetching.

ghost commented:
pgp.mit.edu is working fine for me. The **name:** I fixed because I originally thought they caused the error and checked for the actual string.

conorsch committed:
Prominently documents 30GB disk space requirement

During review of #107, @dachary mentioned running out of disk space. We
used to be better about documenting that! Added to the requirements
section of the README, and cross-referenced from the build guide.

ghost commented:
I ran out of space while having 25GB available. After starting over on a brand new machine, it uses a total of 30GB (system included). I think it's safe to say that if someone has 30GB available it will run successfully.

<pre>
ubuntu@grsec-server:~/ansible-role-grsecurity$ df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            3.8G     0  3.8G   0% /dev
tmpfs           780M  8.8M  771M   2% /run
/dev/sda1        39G   30G  8.8G  78% /
tmpfs           3.9G     0  3.9G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           3.9G     0  3.9G   0% /sys/fs/cgroup
/dev/sda15      105M  3.4M  102M   4% /boot/efi
tmpfs           780M     0  780M   0% /run/user/1000
</pre>

ghost commented:
The build completed successfully following the **Rebuilding the SecureDrop-specific packages from source** instructions
<pre>
vagrant@grsec-rebuild-securedrop:~/linux$ ls -l
total 137200
drwxr-xr-x  3 vagrant vagrant     4096 Jun 21 22:01 etc
-rw-r--r--  1 vagrant vagrant  6762468 Jul 28 09:08 linux-headers-3.14.79-grsec_3.14.79-grsec-10.00.Custom_amd64.deb
-rw-r--r--  1 vagrant vagrant 37782956 Jul 28 09:05 linux-image-3.14.79-grsec_3.14.79-grsec-10.00.Custom_amd64.deb
drwx------ 26 vagrant vagrant     4096 Jul 28 09:18 linux-source-3.14.79-grsec
-rw-r--r--  1 vagrant vagrant 95923178 Jul 28 09:19 linux-source-3.14.79-grsec_3.14.79-grsec-10.00.Custom_all.deb
drwxr-xr-x  9 vagrant vagrant     4096 Jul 28 07:23 ubuntu-package
drwxrwxr-x 27 vagrant vagrant     4096 Jul 28 07:24 ubuntu-trusty
drwxr-xr-x  4 vagrant vagrant     4096 Jun 21 22:01 usr
vagrant@grsec-rebuild-securedrop:~/linux$
</pre>

ghost commented:
Here are intstructions to prepare the build from a pristine ubuntu 17.04, using the libvirt provider (because it is more stable than the virtualbox provider).
<pre>
sudo apt-get update
sudo apt-get install -y vagrant vagrant-libvirt libvirt-bin ansible 
export VAGRANT_DEFAULT_PROVIDER=libvirt
</pre>
it needs a modification of the Vagrant file
<pre>
diff --git a/Vagrantfile b/Vagrantfile
index 840a60f..2cb9122 100644
--- a/Vagrantfile
+++ b/Vagrantfile
@@ -29,7 +29,7 @@ Vagrant.configure("2") do |config|
   # So we'll reuse the package logic from the role to install dependencies,
   # then leave the configuration and build commands to be run manually.
   config.vm.define 'grsec-rebuild-securedrop', autostart: false do |rebuild_sd|
-    rebuild_sd.vm.box = "bento/ubuntu-14.04"
+    rebuild_sd.vm.box = "s3than/trusty64" #"bento/ubuntu-14.04"
     rebuild_sd.vm.hostname = "grsec-rebuild-securedrop"
     rebuild_sd.vm.provision :ansible do |ansible|
       ansible.playbook = 'examples/rebuild-grsecurity-kernel-securedrop.yml'
</pre>
after that, it works as expected.

conorsch commented:
@dachary Excellent, thank you for the careful testing! I've updated the docs recommendation for disk space 20GB -> 30GB. 

> using the libvirt provider (because it is more stable than the virtualbox provider).

So do I! You may find the [vagrant-mutate](https://github.com/sciurus/vagrant-mutate) plugin useful. It allows you to convert any base box to libvirt, which is incredibly helpful when trying to use popular public boxes that don't natively support libvirt/qemu as a backend provider. 

```
vagrant plugin install vagrant-libvirt
vagrant mutate bento/ubuntu-14.04 libvirt
vagrant up --provider libvirt grsec-rebuild-securedrop
```

Currently using `v0.0.40` of vagrant-libvirt (on Debian Stretch) across a number of projects, and have had no trouble with it.

conorsch commented:
@dachary Any further revisions suggested to this PR? If not, I'll merge!

ghost commented:
for the record I peformed 
<pre>
vagrant mutate bento/ubuntu-14.04 libvirt
</pre>
successfully on a pristine strech install and it requires a number of steps to succeed (at least 10). Nothing too complicated but not a oneliner either.
